File: golang/src/Acme/Infrastructure/InMemory/Domain/InMemoryUsers.go
package Domain

import . "../../../Domain/Model"

type InMemoryUsers struct {
	users map[string]*User
}

func NewInMemoryUsers() *InMemoryUsers {
	inMemoryUsers := &InMemoryUsers{}
	inMemoryUsers.users = make(map[string]*User)

	return inMemoryUsers
}

func (class *InMemoryUsers) Add(user *User) {
	class.users[user.Email()] = user
}

func (class *InMemoryUsers) Exists(email string) bool {
	_, exists := class.users[email]

	return exists
}

func (class *InMemoryUsers) GetByEmail(email string) *User {
	return class.users[email]
}

func (class *InMemoryUsers) Remove(email string) {
	delete(class.users, email)
}

File: typescript/src/Acme/Infrastructure/InMemory/Domain/InMemoryUsers.ts
import {Users} from "../../../Domain/Users";
import {User} from "../../../Domain/Model/User";

export class InMemoryUsers implements Users {
    private users: { [key: string]: User; } = {};

    add(user: User): void {
        this.users[user.email] = user;
    }

    exists(email: string): boolean {
        return (email in this.users);
    }

    getByEmail(email: string): User {
        return this.users[email];
    }

    remove(email: string): void {
        delete this.users[email];
    }
}